### DISCUSSION TITLE
Exclude carts with attempted payments from purgeIncompleteCarts

### DISCUSSION THREAD

**Discussion by gopeter** [craftcms/commerce#3684] [Ideas] [answered]
We have a use case for our shop where we have to purge inactive carts really fast, because having an item in the cart reserves an outsourced stock. Like ~15-30 minutes. But the purge removes carts with attempted payments also, and since someone may pay per bank transfer, this cart can't be deleted.

It would be great if we can have a setting to exclude carts with attempted payments from the purge. 

Since attempted payments includes both still-in-progress-payments and failed payments, there maybe has to be a check if the latest transaction status is just "redirect" or "pending", etc. 

If none of this is possible, it would be great if you could introduce an event where I can populate this query: https://github.com/craftcms/commerce/blob/ff3994c915cea65e287bb65f50e85e0d9aca4857/src/services/Carts.php#L418-L422

**Comment by lukeholder** [comment, marked as answer]
Thanks for requesting this. We added this event in 5.3.0.1.

Here is an example of how to modify the query to only target carts with a zero total price.

```php
use craft\commerce\events\CartPurgeEvent;
use craft\commerce\services\Carts;
use yii\base\Event;

Event::on(
  Carts::class,
  Carts::EVENT_BEFORE_PURGE_INACTIVE_CARTS,
  function(CartPurgeEvent $event) {
      $event->inactiveCartsQuery = $event->inactiveCartsQuery->andWhere(['totalPrice' => 0]);
  }
);
```

Thanks!
